src/combinatorics/ssyt.lean

```diff
@@ -0,0 +1,85 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import combinatorics.young_diagram
+
+/-!
+# Semistandard Young tableaux
+
+A semistandard Young tableau is a filling of a Young diagram by natural numbers, such that
+the entries are weakly increasing left-to-right along rows (i.e. for fixed `i`), and
+strictly-increasing top-to-bottom along columns (i.e. for fixed `j`).
+
+An example of an SSYT of shape `μ = [4, 2, 1]` is:
+
+```text
+0 0 0 2
+1 1
+2
+```
+
+We represent an SSYT as a function `ℕ → ℕ → ℕ`, which is required to be zero for all pairs
+`(i, j) ∉ μ` and to satisfy the row-weak and column-strict conditions on `μ`.
+
+
+## Main definitions
+
+- `ssyt (μ : young_diagram)` : semistandard Young tableaux of shape μ
+- `ssyt.highest_weight (μ : young_diagram)`: the semistandard Young tableau whose `i`th row
+  consists entirely of `i`s, for each `i`.
+
+## Tags
+
+Semistandard Young tableau
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+/-- A semistandard Young tableau (SSYT) is a filling of the cells of a Young diagram by natural
+numbers, such that the entries in each row are weakly increasing (left to right), and the entries
+in each column are strictly increasing (top to bottom).
+
+Here, an SSYT is represented as an unrestricted function `ℕ → ℕ → ℕ`, but which is required to
+vanish outside `μ`. --/
+@[ext] structure ssyt (μ : young_diagram) :=
+  (entry : ℕ → ℕ → ℕ)
+  (row_weak : ∀ {i j1 j2 : ℕ}, j1 < j2 → (i, j2) ∈ μ → entry i j1 ≤ entry i j2)
+  (col_strict : ∀ {i1 i2 j: ℕ}, i1 < i2 → (i2, j) ∈ μ → entry i1 j < entry i2 j)
+  (zeros' : ∀ {i j}, (i, j) ∉ μ → entry i j = 0)
+
+namespace ssyt
+
+instance {μ : young_diagram} : has_coe_to_fun (ssyt μ) (λ _, ℕ → ℕ → ℕ) :=
+{ coe := λ T i j, T.entry i j }
+
+@[simp] lemma to_entry {μ : young_diagram} (T : ssyt μ) (i j : ℕ) :
+  T.entry i j = T i j := rfl
+
+lemma zeros {μ : young_diagram} (T : ssyt μ)
+  {i j : ℕ} (not_cell : (i, j) ∉ μ) : T i j = 0 := T.zeros' not_cell
+
+lemma row_weak_of_le {μ : young_diagram} (T : ssyt μ) {i j1 j2 : ℕ}
+  (hj : j1 ≤ j2) (cell : (i, j2) ∈ μ) : T i j1 ≤ T i j2 :=
+by {cases eq_or_lt_of_le hj, subst h, exact T.row_weak h cell}
+
+lemma col_weak {μ : young_diagram} (T : ssyt μ) {i1 i2 j : ℕ}
+  (hi : i1 ≤ i2) (cell : (i2, j) ∈ μ) : T i1 j ≤ T i2 j :=
+by {cases eq_or_lt_of_le hi, subst h, exact le_of_lt (T.col_strict h cell)}
+
+/-- The "highest weight" SSYT of a given shape is has all i's in row i, for each i. --/
+def highest_weight (μ : young_diagram) : ssyt μ :=
+{ entry := λ i j, ite ((i, j) ∈ μ) i 0,
```

Comment:
You're free to use `ite`, but I find its notation to be more legible personally:
```suggestion
{ entry := λ i j, if (i, j) ∈ μ then i else 0,
```